huggingface_pipelines/text.py

```diff
@@ -0,0 +1,156 @@
+        logger.info("Processing batch...")
+        texts = batch['text']
+        embeddings = self.encode_texts(texts)
+        reconstructed_texts = self.decode_embeddings(embeddings)
+        bleu_score = self.compute_bleu(texts, reconstructed_texts)
+        logger.info(f"Batch processed. BLEU score: {bleu_score}")
+        return {'original': texts, 'reconstructed': reconstructed_texts, 'bleu': bleu_score}
+
+    def process_batches(self):
+        """Processes all batches in the dataset and stores the results."""
+        try:
+            logger.info("Starting to process batches...")
+            # Limit to the first 1000 examples
+            dataset = self.dataset.select(range(min(len(self.dataset), 1000)))
+
+            # Process the dataset using map
+            results = dataset.map(
+                lambda batch: self.process_batch(batch),
+                batched=True,
+                batch_size=self.batch_size,
+                remove_columns=dataset.column_names,
+                load_from_cache_file=False
+            )
+            self.results.extend([{k: v[i] for k, v in results.items()}
+                                for i in range(len(results[next(iter(results))]))])
+
+            logger.info("All batches processed. Caching results...")
+            self.cache_results()
+            logger.info("Results cached successfully.")
+        except Exception as e:
+            logger.error(f"Error processing batches: {e}")
+
+    def cache_results(self):
+        """Caches the results to a JSON file."""
+        try:
+            logger.info("Caching results to results.json...")
+            with open('results.json', 'w') as f:
+                json.dump(self.results, f)
+            logger.info("Results cached successfully.")
+        except Exception as e:
+            logger.error(f"Error caching results: {e}")
+
+    def analyze_results(self):
+        """Analyzes the results to determine the percentage of batches with low BLEU scores."""
+        if not self.results:
+            logger.warning("No results to analyze.")
+            return
+
+        logger.info("Analyzing results...")
+        low_bleu_count = sum(
+            1 for result in self.results if result['bleu'][0] < self.low_bleu_threshold)
+        total_batches = len(self.results)
+        low_bleu_percentage = (low_bleu_count / total_batches) * 100
+
+        logger.info(
+            f"Percentage of batches with BLEU score below {self.low_bleu_threshold}: {low_bleu_percentage:.2f}%")
+        self.report_low_bleu_scores()
+
+    def report_low_bleu_scores(self):
+        """Reports batches with BLEU scores below the threshold."""
+        for result in self.results:
+            if result['bleu'][0] < self.low_bleu_threshold:
+                logger.info(f"Low BLEU score detected: {result['bleu'][0]}")
+                logger.info(f"Original Text: {result['original']}")
+                logger.info(f"Reconstructed Text: {result['reconstructed']}")
```

Comment:
This should be a part of BleuAnalyser class and not of PipelineFactory. BleuAnalyser class can have its own config and can be called on a `hf_sonarized_dataset` to get the quality results.